### PR TITLE
feat(metrics_indexer): Add writes limiter + fixes to sliding windows rate limiter [INGEST-1380]

### PR DESCRIPTION
**Pull request by untitaker**
Split from #36531 

**Review https://github.com/getsentry/sentry/pull/37171 first!**

This introduces the writes limiter to the codebase and contains related fixes to the sliding window rate limiter found while unittesting. It does not integrate the writes limiter into postgres yet.

### CODE REVIEW

**Review by onewland — APPROVED** [review]

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+
+@dataclasses.dataclass
+class DroppedString:
+    key_result: KeyResult
+    fetch_type: FetchType
+    fetch_type_ext: FetchTypeExt
+
+
+class WritesLimiter:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:96]: Why not making this a context manager? 
Then you would not have to give back the state to the code that writes on postgres and relying on that to pass it back to the rate limtier to `apply_write_limits`.

Example is how we did it in snuba 
https://github.com/getsentry/snuba/blob/master/snuba/state/rate_limit.py#L94

You can yield the list of strings to write and those to drop. 
Personally I would make it a class and not a generator, but this is just personal preference.

File: src/sentry/sentry_metrics/indexer/ratelimiters.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        org_ids, requests = _construct_quota_requests(use_case_id, keys)
+        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+
+        granted_key_collection = dict(keys.mapping)
+        dropped_strings = []
+
+        for org_id, request, grant in zip(org_ids, requests, grants):
+            allowed_strings = granted_key_collection[org_id]
+            if len(allowed_strings) > grant.granted:
+                allowed_strings = set(allowed_strings)
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/ratelimiters.py:134]: I know that `allowed_strings` and `granted_key_collection` are mutated through this loop and, in the end they represent what their names say, but that makes this code quite hard to follow as when reading it two variables do not mean what their name says till the end of the loop.

I would suggest this: 
- you start with an empty `allowed_strings` and `dropped_strings`
- take `grant.granted` strings from `granted_key_collection` and put them into `allowed_strings`, the rest goes in `dropped`.

The iterations you skip by mutating `allowed_strings` in place is likely to be irrelevant considering that in the same code path we have done a redis `mget` before getting here.

**Review by fpacifici — COMMENTED** [review]
Up to you if you want to implement the changes suggested in comments in line

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:115]: done. I made the `state` itself a context manager

File: src/sentry/sentry_metrics/indexer/ratelimiters.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:150]: I'm going to roll this back.. I tried to fix the testfailures, and I think this just causes more edgecases to be handled in case there are no quotas configured.

Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:150]: I think you would be doing the same amount of clone operations with my suggestion that you are doing now. But again, I don't think it matters as in the same code path (when we got here) we did an mget from redis which is likely to dwarf that impact. 

This is not so critical as it is not in the code path taken by the majority of the lookup operations. THat code path lands on memcache and does not get here.

Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/ratelimiters.py:150]: sry @fpacifici, i overlooked this. if you don't mind i will punt on this and see if it becomes a problem again.. i think i painted myself a bit into a corner with the way i am passing data around as well, so i think in order to make this work i would have to change some other signatures as well.
